src/Features/CSharp/Portable/InlineParameterNameHints/InlineParameterNameHintsService.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.InlineParameterNameHints;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp.InlineParameterNameHints
+{
+    /// <summary>
+    /// The service to locate the positions in which the adornments should appear
+    /// as well as associate the adornments back to the parameter name
+    /// </summary>
+    [ExportLanguageService(typeof(IInlineParameterNameHintsService), LanguageNames.CSharp), Shared]
+    internal class InlineParameterNameHintsService : IInlineParameterNameHintsService
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public InlineParameterNameHintsService()
+        {
+        }
+
+        public async Task<IEnumerable<InlineParameterHint>> GetInlineParameterNameHintsAsync(
+            Document document,
+            TextSpan textSpan,
+            CancellationToken cancellationToken)
+        {
+            var tree = await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var node = await tree.GetRootAsync(cancellationToken).ConfigureAwait(false);
+            var spans = new List<InlineParameterHint>();
+
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var invocations = node.DescendantNodes(textSpan).OfType<InvocationExpressionSyntax>();
```

Comment:
Rather than calling DescendantNodes twice, let's only call it once, and instead inside the foreach check for InvocationExpressionSyntax and AttributeListSyntax; that should be a bit more efficient on the syntax trees and might avoid realizing things.

Reply:
Fixed.